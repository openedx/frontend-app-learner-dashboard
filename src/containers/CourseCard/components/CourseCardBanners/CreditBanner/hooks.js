import { StrictDict } from '../../../../../utils';
import { reduxHooks } from '../../../../../hooks';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = reduxHooks.useCardCreditData(cardId);
  const { supportEmail } = reduxHooks.usePlatformSettingsData();
  if (!credit.isEligible) {
    return null;
  }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
