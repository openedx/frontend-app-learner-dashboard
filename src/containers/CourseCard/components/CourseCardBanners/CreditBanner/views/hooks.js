import React from 'react';

import { StrictDict } from '@src/utils';
import { apiHooks } from '@src/hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest()
      .then((request) => {
        setRequestData(request.data);
      });
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
