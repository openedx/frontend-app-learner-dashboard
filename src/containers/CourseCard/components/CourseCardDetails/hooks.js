import { useIntl } from '@edx/frontend-platform/i18n';
import { utilHooks, reduxHooks } from 'hooks';

import * as hooks from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const enrollment = reduxHooks.useCardEnrollmentData(cardId);
  const courseRun = reduxHooks.useCardCourseRunData(cardId);
  const formatDate = utilHooks.useFormatDate();
  if (!courseRun.isStarted) {
    if (!courseRun.startDate && !courseRun.advertisedStart) { return null; }
    const startDate = courseRun.advertisedStart ? courseRun.advertisedStart : formatDate(courseRun.startDate);
    return formatMessage(messages.courseStarts, { startDate });
  }
  if (enrollment.isEnrolled) {
    const { isArchived, endDate } = courseRun;
    const {
      accessExpirationDate,
      isAuditAccessExpired,
    } = enrollment;

    if (enrollment.isAudit && accessExpirationDate) {
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    if (!endDate) { return null; }
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = reduxHooks.useCardProviderData(cardId).name;
  const { courseNumber } = reduxHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = reduxHooks.useCardEntitlementData(cardId);

  const openSessionModal = reduxHooks.useUpdateSelectSessionModalCallback(cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: hooks.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    courseNumber,
    changeOrLeaveSessionMessage: formatMessage(messages.changeOrLeaveSessionButton),
  };
};

export default useCardDetailsData;
