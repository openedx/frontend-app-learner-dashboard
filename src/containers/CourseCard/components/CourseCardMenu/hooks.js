import { useKeyedState } from '@edx/react-unit-test-utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

export const stateKeys = {
  isUnenrollConfirmVisible: 'isUnenrollConfirmVisible',
  isEmailSettingsVisible: 'isEmailSettingsVisible',
};

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = useKeyedState(stateKeys.isUnenrollConfirmVisible, false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = useKeyedState(stateKeys.isEmailSettingsVisible, false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(
    track.course.courseOptionsDropdownClicked,
    cardId,
  );
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useOptionVisibility = (cardId) => {
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);

  const shouldShowUnenrollItem = isEnrolled && !isEarned;
  const shouldShowDropdown = (
    shouldShowUnenrollItem
    || isEmailEnabled
    || facebook.isEnabled
    || twitter.isEnabled
  );

  return {
    shouldShowUnenrollItem,
    shouldShowDropdown,
  };
};
