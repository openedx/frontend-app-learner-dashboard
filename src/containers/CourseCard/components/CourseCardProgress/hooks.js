import { useState, useEffect } from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';

import { reduxHooks } from 'hooks';
import messages from './messages';

export const useCourseProgress = ({ cardId }) => {
  const [progress, setProgress] = useState(0);
  const [completeCount, setCompleteCount] = useState(0);
  const [totalCount, setTotalCount] = useState(0);
  const [loading, setLoading] = useState(false);
  const { isEnrolled, hasStarted } = reduxHooks.useCardEnrollmentData(cardId);
  const { courseId } = reduxHooks.useCardCourseRunData(cardId);

  useEffect(() => {
    const fetchProgress = async () => {
      if (!isEnrolled || !hasStarted || !courseId) {
        setProgress(0);
        setCompleteCount(0);
        setTotalCount(0);
        return;
      }

      setLoading(true);
      try {
        const client = getAuthenticatedHttpClient();
        const lmsApiUrl = `${getConfig().LMS_BASE_URL}/api/course_home/v1/progress/${courseId}`;
        const response = await client.get(lmsApiUrl);
        const progressData = response.data;

        let progressPercent = 0;
        let complete = 0;
        let total = 0;

        // Check for completion_summary from course_home API
        if (progressData.completion_summary) {
          const { complete_count, incomplete_count, locked_count } = progressData.completion_summary;
          complete = complete_count;
          total = complete_count + incomplete_count + locked_count;
          if (total > 0) {
            progressPercent = Math.round((complete_count / total) * 100);
          }
        } else if (progressData.completion_percentage !== undefined) {
          progressPercent = progressData.completion_percentage;
        } else if (progressData.progress?.percent !== undefined) {
          progressPercent = Math.round(progressData.progress.percent * 100);
        } else if (progressData.progress_percentage !== undefined) {
          progressPercent = progressData.progress_percentage;
        } else if (progressData.percent !== undefined) {
          progressPercent = progressData.percent;
        }

        setProgress(Math.max(0, Math.min(100, progressPercent)));
        setCompleteCount(complete);
        setTotalCount(total);
      } catch (error) {
        console.warn('Failed to fetch course progress:', error);
        setProgress(0);
        setCompleteCount(0);
        setTotalCount(0);
      } finally {
        setLoading(false);
      }
    };

    fetchProgress();
  }, [cardId, isEnrolled, hasStarted, courseId]);

  return { progress, completeCount, totalCount, loading };
};

export const useProgressVariant = ({ progress }) => {
  if (progress >= 75) return 'success';
  if (progress >= 50) return 'warning';
  return 'info';
};

export const useCardProgressData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const { isEnrolled } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    progress, completeCount, totalCount, loading,
  } = useCourseProgress({ cardId });
  const variant = useProgressVariant({ progress });

  return {
    shouldRender: isEnrolled,
    progress,
    completeCount,
    totalCount,
    loading,
    variant,
    progressLabel: formatMessage(messages.courseProgress),
    loadingLabel: formatMessage(messages.loading),
  };
};

export default useCardProgressData;
