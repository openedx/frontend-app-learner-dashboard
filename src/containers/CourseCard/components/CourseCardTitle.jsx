import React from 'react';
import PropTypes from 'prop-types';

import track from '@src/tracking';
import { reduxHooks } from '@src/hooks';
import useActionDisabledState from './hooks';

const { courseTitleClicked } = track.course;

export const CourseCardTitle = ({ cardId }) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { homeUrl } = reduxHooks.useCardCourseRunData(cardId);
  const handleTitleClicked = reduxHooks.useTrackCourseEvent(
    courseTitleClicked,
    cardId,
    homeUrl,
  );
  const { disableCourseTitle } = useActionDisabledState(cardId);
  return (
    <h3>
      {disableCourseTitle ? (
        <span className="course-card-title" data-testid="CourseCardTitle">{courseName}</span>
      ) : (
        <a
          href={homeUrl}
          className="course-card-title"
          data-testid="CourseCardTitle"
          onClick={handleTitleClicked}
        >
          {courseName}
        </a>
      )}
    </h3>
  );
};

CourseCardTitle.propTypes = {
  cardId: PropTypes.string.isRequired,
};

CourseCardTitle.defaultProps = {};

export default CourseCardTitle;
