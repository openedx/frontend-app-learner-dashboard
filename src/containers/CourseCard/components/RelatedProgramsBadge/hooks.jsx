import React from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from '@src/utils';
import { reduxHooks } from '@src/hooks';

import messages from './messages';
import * as module from './hooks';

export const state = StrictDict({
  isOpen: (val) => React.useState(val), // eslint-disable-line
});

export const useRelatedProgramsBadgeData = ({ cardId }) => {
  const [isOpen, setIsOpen] = module.state.isOpen(false);
  const { formatMessage } = useIntl();
  const numPrograms = reduxHooks.useCardRelatedProgramsData(cardId).length;
  let programsMessage = '';
  if (numPrograms) {
    programsMessage = formatMessage(
      numPrograms === 1 ? messages.badgeLabelSingular : messages.badgeLabelPlural,
      { numPrograms },
    );
  }

  return {
    numPrograms,
    programsMessage,
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};

export default useRelatedProgramsBadgeData;
