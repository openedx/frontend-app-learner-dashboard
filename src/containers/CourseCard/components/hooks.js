import { reduxHooks } from 'hooks';

export const useActionDisabledState = (cardId) => {
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const {
    hasAccess, isAudit, isAuditAccessExpired,
  } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    isEntitlement, isFulfilled, canChange, hasSessions,
  } = reduxHooks.useCardEntitlementData(cardId);

  const { resumeUrl, homeUrl } = reduxHooks.useCardCourseRunData(cardId);

  const disableBeginCourse = !homeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableResumeCourse = !resumeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableViewCourse = !hasAccess || (isAudit && isAuditAccessExpired);
  const disableSelectSession = !isEntitlement || isMasquerading || !hasAccess || (!canChange || !hasSessions);

  const disableCourseTitle = (isEntitlement && !isFulfilled) || disableViewCourse;

  return {
    disableBeginCourse,
    disableResumeCourse,
    disableViewCourse,
    disableSelectSession,
    disableCourseTitle,
  };
};

export default useActionDisabledState;
