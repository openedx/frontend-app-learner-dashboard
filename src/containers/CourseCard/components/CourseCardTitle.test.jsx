import { shallow } from '@edx/react-unit-test-utils';

import { reduxHooks } from 'hooks';
import track from 'tracking';
import useActionDisabledState from './hooks';
import CourseCardTitle from './CourseCardTitle';

const homeUrl = 'home-url';

jest.mock('tracking', () => ({
  course: {
    courseTitleClicked: jest.fn().mockName('segment.courseTitleClicked'),
  },
}));

jest.mock('hooks', () => ({
  reduxHooks: {
    useCardCourseData: jest.fn(() => ({ courseName: 'course-name' })),
    useCardCourseRunData: jest.fn(() => ({ homeUrl })),
    useTrackCourseEvent: jest.fn((eventName, cardId, url) => ({
      trackCourseEvent: { eventName, cardId, url },
    })),
  },
}));
jest.mock('./hooks', () => jest.fn(() => ({ disableCourseTitle: false })));

describe('CourseCardTitle', () => {
  const props = {
    cardId: 'cardId',
  };
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('snapshot', () => {
    test('renders clickable link course title', () => {
      const wrapper = shallow(<CourseCardTitle {...props} />);
      expect(wrapper.snapshot).toMatchSnapshot();
      const title = wrapper.instance.findByTestId('CourseCardTitle');
      expect(title[0].type).toBe('a');
      expect(title[0].props.onClick).toEqual(
        reduxHooks.useTrackCourseEvent(
          track.course.courseTitleClicked,
          props.cardId,
          homeUrl,
        ),
      );
    });
    test('renders disabled link', () => {
      useActionDisabledState.mockReturnValueOnce({ disableCourseTitle: true });
      const wrapper = shallow(<CourseCardTitle {...props} />);
      expect(wrapper.snapshot).toMatchSnapshot();
      const title = wrapper.instance.findByTestId('CourseCardTitle');
      expect(title[0].type).toBe('span');
      expect(title[0].props.onClick).toBeUndefined();
    });
  });
  describe('behavior', () => {
    it('initializes', () => {
      shallow(<CourseCardTitle {...props} />);
      expect(reduxHooks.useCardCourseData).toHaveBeenCalledWith(props.cardId);
      expect(reduxHooks.useCardCourseRunData).toHaveBeenCalledWith(
        props.cardId,
      );
      expect(useActionDisabledState).toHaveBeenCalledWith(props.cardId);
    });
  });
});
