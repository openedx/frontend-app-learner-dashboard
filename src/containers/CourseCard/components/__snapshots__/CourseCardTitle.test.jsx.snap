// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCardTitle snapshot renders clickable link course title 1`] = `
<h3>
  <a
    className="course-card-title"
    data-testid="CourseCardTitle"
    href="home-url"
    onClick={
      {
        "trackCourseEvent": {
          "cardId": "cardId",
          "eventName": [MockFunction segment.courseTitleClicked],
          "url": "home-url",
        },
      }
    }
  >
    course-name
  </a>
</h3>
`;

exports[`CourseCardTitle snapshot renders disabled link 1`] = `
<h3>
  <span
    className="course-card-title"
    data-testid="CourseCardTitle"
  >
    course-name
  </span>
</h3>
`;
